### PR TITLE
SONARJAVA-1767 Calling method get() before isPresent() on Optional check

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+  }
+
+  @Override
+  public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
+    final PreStatementVisitor visitor = new PreStatementVisitor(context);
+    syntaxNode.accept(visitor);
+    return visitor.programState;
+  }
+
+  private static class OptionalSymbolicValue extends SymbolicValue {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java]: MAJOR Override this superclass' "equals" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2160)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import java.util.List;
+
+@Rule(key = "S3655")
+public class OptionalGetBeforeIsPresentCheck extends SECheck {
+
+  private static final String JAVA_UTIL_OPTIONAL = "java.util.Optional";
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java]: constants should be scoped in prestatement visitor

File: java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    /**
+     * Will be called only after calling optional.isPresent()
+     */
+    @Override
+    public List<ProgramState> setConstraint(ProgramState programState, BooleanConstraint booleanConstraint) {
+      ObjectConstraint optionalConstraint = (ObjectConstraint) programState.getConstraint(optionalSV);
+      if (optionalConstraint == null) {
+        return ImmutableList.of();
+      }
+      boolean isknownAsNotPresent = optionalConstraint.hasStatus(Status.NOT_PRESENT);
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java]: Seems like only one status could be tested.

**Pull request merged**